test/metabase/models/setting_test.clj

```diff
@@ -1204,45 +1204,189 @@
               #"^Surprise!$"
               (#'setting/realize x)))))))
 
-(deftest valid-json-setting-test
-  (mt/with-temp-env-var-value ["MB_TEST_JSON_SETTING" "[1, 2]"]
-    (is (nil? (setting/validate-settings-formatting!)))))
-
-(defn- get-parse-exception [raw-value]
-  (mt/with-temp-env-var-value ["MB_TEST_JSON_SETTING" raw-value]
-    (try
-      (setting/validate-settings-formatting!)
-      (throw (java.lang.RuntimeException. "This code should never be reached."))
-      (catch java.lang.Exception e e))))
-
-(defn- assert-parser-exception! [ex cause-message]
-  (is (= "Invalid JSON configuration for setting: test-json-setting" (ex-message ex)))
+
+(defmacro define-setting-for-type [format]
+  `(defsetting ~(symbol (str "test-" (name format) "-setting"))
+     "Setting to test validation of this format - this only shows up in dev"
+     :type ~(keyword (name format))))
+
+(defmacro define-settings-for-types [& formats]
+  (list 'dorun (cons 'list (for [f formats] `(define-setting-for-type ~f)))))
+
+;; Create stub settings for the remaining types that weren't used by any previous tests
+(define-settings-for-types :double :keyword :integer :positive-integer :timestamp)
+
+(defn- get-formatting-exception []
+  (try
+    (setting/validate-settings-formatting!)
+    nil
+    (catch java.lang.Exception e e)) )
+
+(defn- get-json-parse-exception [raw-value]
+  (mt/with-temp-env-var-value [mb-test-json-setting raw-value]
+    (get-formatting-exception)))
+
+(defn- get-csv-parse-exception [raw-value]
+  (mt/with-temp-env-var-value [mb-test-csv-setting raw-value]
+    (get-formatting-exception)))
```

Comment:
I could double down on drying things up with macros, but I thought it was nice for my IDE to know about these symbols. I could go in the other direction and reduce the tests all the way and make the tests completely macro driven, e.g.

```clojure 
(assert-valid :json "eg1" "eg2" ...)

(assert-invalid :json 
  "blag" #"Exception message pattern"
  ...)
  ```

Reply:
You could DRY up these so you can write `(get-parse-exception mb-test-csv-setting)` in the test, and that would be tidier to me. I think the limit would be removing the assertions from the tests. It might be fewer characters but it doesn't add much readability IMO.

Reply:
Nice suggestion, I think I can riff off that.